draft-yasskin-dispatch-bundled-exchanges.md

```diff
@@ -0,0 +1,436 @@
+
+Load a CBOR item starting at *index-start*. Parsing MUST fail if any of the
+following is true:
+
+* This item doesn't match the `index` rule in the following CDDL
+  ({{!I-D.ietf-cbor-cddl}}).
+* This item isn't encoded canonically ({{canonical-encoding}}).
+* Any key is not a valid request item ({{request-items}}).
+
+~~~ cddl
+index = {* http-request => [ offset: uint,
+                             length: uint] }
+~~~
+
+Adjust all of the offsets by adding *responses-start*, so they point at the
+start of the response relative to the start of the whole bundle. Note that
+offsets are stored relative to the responses section instead of the start of the
+whole bundle so that they don't need to incorporate their own size, which would
+make it more difficult to encode them canonically.
+
+Interpret the keys as requests as described in {{request-items}}, and return
+this set of requests with each request's adjusted offset and length attached.
+
+### Parsing the manifest section {#manifest-section}
+
+To parse the manifest section, starting at offset *manifest-start*, the parser
+MUST load a CBOR item starting at *manifest-start*. If this item isn't a valid
+request item, as defined by {{request-items}}, the parser MUST fail.
+
+~~~ cddl
+manifest = http-request
+~~~
+
+Return the request represented by this parsed item, interpreted as defined by
+{{request-items}}.
+
+### Parsing the critical section {#critical-section}
+
+To parse the critical section, starting at offset *critical-start*, the parser
+MUST load a CBOR item starting at *critical-start*. If this item doesn't match
+the `critical` rule in the following CDDL ({{!I-D.ietf-cbor-cddl}}), the parser
+MUST fail.
+
+~~~ cddl
+critical = [*tstr]
+~~~
+
+For each value *section-name* in the `critical` list, if the parser does not
+support sections named *section-name*, the parser MUST fail.
+
+This section returns no information.
+
+### Interpreting request CBOR items {#request-items}
+
+HTTP requests are represented by CBOR items matching the following CDDL
+({{!I-D.ietf-cbor-cddl}}).
+
+~~~ cddl
+http-request = {
+  * bstr => bstr
+}
+~~~
+
+A valid request item *R*:
+
+* Matches the above CDDL.
+* Has only lower-case ASCII keys, matching the requirement in Section 8.1.2 of
+  {{?RFC7540}}.
+* Has exactly two keys starting with a ':' character, ':method' and ':url'.
+* *R*\[':method'] is an HTTP method defined as cacheable (Section 4.2.3 of
+  {{!RFC7231}}) and safe (Section 4.2.1 of {{!RFC7231}}), as required for
+  PUSH_PROMISEd requests (Section 8.2 of {{?RFC7540}}).
+* *R*\[':url'] is an absolute URI (Section 4.3 of {{!RFC3986}}).
+
+A valid request item *R* is interpreted as an HTTP request by interpreting
+*R*\[':method'] as the request's method (Section 4 of {{!RFC7231}}),
+*R*\[':url'] as the request's effective request URI (Section 5.5 of
+{{!RFC7230}}), and the remaining key/value pairs as the request's header fields.
```

Comment:
Those key/value pairs are suitably constrained somehow? What's the parsing for them?

Reply:
They're HTTP headers, so we should do whatever https://fetch.spec.whatwg.org/#http-network-fetch does.

Or are you worried about how claimed request headers are matched against browser request headers? That's not written yet. I was hoping to use the matching rules for PUSH_PROMISEs, but those aren't written yet either.

Reply:
Well they're not quite the same as ordinary HTTP headers, since they are delimited differently, right? So e.g., can you have a 0x0A byte in the value?

Reply:
HTTP/2 headers are also delimited in a way that lets their values include a 0x0A byte. I wonder if browsers handle that correctly...

I suspect the right thing to do here is to explicitly create a [header list](https://fetch.spec.whatwg.org/#concept-header-list) from this, but since that concept doesn't exist in the IETF, I'll have to give up on the idea of making this an IETF draft.

I'm going to ping some folks to see if I'll hurt any feelings by doing that.

Reply:
Oooh wow, I didn't know that about H/2. That's another reason to get a web-platform-tests setup.